Implement multiply-add interface in LinearAlgebra

tkf opened the pull request:
I started working on "multiply-add" interface for computing _C = αAB + βC_  for `LinearAlgebra`. The detail of the API is currently discussed in JuliaLang/LinearAlgebra.jl#473.  I go ahead and implement the feature using the name `addmul!`.  I use this name only because it's easy to switch to other alternatives by simple replace.  Let's do (and finish :smile:) the naming discussion in JuliaLang/LinearAlgebra.jl#473.

Remaining TODOs:

* [x] Implement multiply-add interface for most (?) of the types in LinearAlgebra
* [x] Pass existing tests for `mul!`
* [x] Tests for matmul.jl
* [x] Tests for generic.jl
* [x] Tests for uniformscaling.jl
* [x] Tests for diagonal.jl
* [x] Tests for bidiag.jl
* [x] Tests for triangular.jl
* [x] Benchmark / optimization
* [x] Pick a final name -> `mul!` https://github.com/JuliaLang/LinearAlgebra.jl/issues/473
- [ ] decide and implement NaN behavior

EDIT: I updated the summary "inplace" since I noticed there were some now irrelevant comments. (You can click `edited ▼` above to see the old version.)

tkf committed:
Multiply-add interface for BLAS.herk!

tkf committed:
Multiply-add interface for BLAS.syrk!

tkf committed:
Multiply-add interface for gemv!

tkf committed:
Fix UndefRefError from C[i,j]

It may not be defined for Matrix{BigFloat}.

tkf committed:
Multiply-add interface for BLAS.gemm!

tkf committed:
Do not assume *(::Bool, ::eltype(C)) exists

jebej commented:
Wouldn't the two functions defined do exactly the same thing for the default arguments of `alpha` and `beta` (which should be `true` and `false`) since `gemm_wrapper!` already has those defaults? If so, what would be the point, except for those two functions to have a different name?

tkf commented:
The idea is that ternary `mul!` and `muladd!` do `C = A * B` and `C += A * B`, respectively, to make `muladd!` similar to scalar `muladd`.  Of course, the exact signature may be different from the one I write above.  Maybe only 5-ary `muladd!` will be defined.  Let's keep the API discussion in #23919.

tkf committed:
Test multiply-add interface

tkf committed:
Implement mul! in terms of addmul!

tkf committed:
Document addmul!

tkf committed:
Add multiply-add interface for symmetric matrices

tkf committed:
Add multiply-add interface for Number and UniformScaling

tkf committed:
Use lmul! for beta * C; eltype may not be commutative

tkf committed:
Add multiply-add interface for diagonal matrices

tkf committed:
Add _lmul_or_fill!

tkf committed:
Add multiply-add interface for bi- and tri-diagonal matrices

tkf committed:
Add multiply-add interface for triangular matrices

tkf committed:
Test multiply-add interface in test/generic.jl

tkf committed:
Fix addmul!(C, s::Number, X, alpha, beta)

and addmul!(C, X, s::Number, alpha, beta)

tkf committed:
Special-case alpha=1 beta=0 using type parameter

tkf committed:
Test multiply-add interface in test/uniformscaling.jl

tkf committed:
Use addmul! in SparseArrays

tkf committed:
Test multiply-add interface in test/diagonal.jl

tkf committed:
Systematically test addmul!

tkf committed:
Make MulAddMul benchmark-friendly

tkf committed:
Use MulAddMul in A_mul_B_td!

tkf committed:
Fix _modify! docstring

tkf committed:
Comment out broken test_broken

tkf committed:
Relax rtol based on eltype of matrices A, B, C

tkf commented on stdlib/LinearAlgebra/test/addmul.jl:
```diff
@@ -0,0 +1,196 @@
+end
+
+n_samples = 1.5
+# n_samples = Inf  # to try all combinations
```
Note: `n_samples = Inf` takes about 4 hours to finish

tkf committed:
Pass around MulAddMul instead of alpha and beta for type stability

tkf committed:
Inline functions between *(::Matrix, ::Matrix) and gemm_wrapper!

This is required for recovering the performance of current master.
Checked with:

    A = [1. 0; 2 0]
    @﻿benchmark $A * $A

tkf commented:
I think this PR is ready for review.  I'm working on optimizations but I'd like to know if the direction is OK before optimizing the current code too much.

I think the most potentially controversial point is the type I introduced and used everywhere in `addmul!` definitions:

https://github.com/JuliaLang/julia/blob/96c5550bf78cbcc3a7489c45b361a4d500cf51c0/stdlib/LinearAlgebra/src/matmul.jl#L5-L22

The idea is to specialize the method for `addmul!` internal function (e.g., `generic_matmatmul!`) for the case `alpha = 1` and `beta = 0` so that we don't pay the cost at run-time.  This is done simply by avoiding unnecessary arithmetics:

https://github.com/JuliaLang/julia/blob/96c5550bf78cbcc3a7489c45b361a4d500cf51c0/stdlib/LinearAlgebra/src/matmul.jl#L27-L30

(Note that there is an additional helper function `_modify!` because we need to support `BigFloat` etc.)

Here is a benchmark supports that it's effective:

```julia
using LinearAlgebra
using LinearAlgebra: _addmul!, MulAddMul

N = 100
A = UpperTriangular(randn(N, N))
b = randn()
C = copy(A)
_add1 = MulAddMul(1, 0)
_add2 = MulAddMul{false, false, Int, Int}(1, 0)  # pretend alpha!=1, beta!=0

using BenchmarkTools
(b1 = @benchmark _addmul!($C, $A, $b, $_add1) setup=(fill!($C, 0))) |> display
(b2 = @benchmark _addmul!($C, $A, $b, $_add2) setup=(fill!($C, 0))) |> display

judge(minimum(b1), minimum(b2))
```

which says avoiding additional arithmetics makes `mul!` ~25% faster (which is a trivial statement but I thought I'd better check the implementation anyway):

```
BenchmarkTools.TrialJudgement:
  time:   -24.93% => improvement (5.00% tolerance)
  memory: +0.00% => invariant (1.00% tolerance)
```

tkf commented:
Failure in macOS on Travis seems to be an irrelevant build issue https://travis-ci.org/JuliaLang/julia/builds/457691742

tkf commented:
To recover the performance of `*(::Matrix, ::Matrix)` for 2x2 matrices, I had to inline `mul!` and `addmul!` (which are the functions between `*(::Matrix, ::Matrix)` and `gemm_wrapper!`) 46d66149bde3f651b80f607c1893e68332100b59.  I tried invoking `addmul!` directly from `*` but it didn't change the performance.

I only changed `addmul!` for single definition relevant to this benchmark.  So I think there may be regressions for other matrix types.  Shall I put `@inline` for all `addmul!`?

---

At master ff74f6bc77e42c974b0f890ff25701f38f362560:

```
julia> Base.GIT_VERSION_INFO.branch, Base.GIT_VERSION_INFO.commit
("master", "ff74f6bc77e42c974b0f890ff25701f38f362560")

julia> using BenchmarkTools
       A = [1. 0; 2 0];
       @btime $A * $A;
  32.406 ns (1 allocation: 112 bytes)
```

Before 46d66149bde3f651b80f607c1893e68332100b59:

```
julia> Base.GIT_VERSION_INFO.branch, Base.GIT_VERSION_INFO.commit
("matmuladd", "57a60772316e6d58ab11b5d9bb327bf80cf9ea02")

julia> using BenchmarkTools
       A = [1. 0; 2 0];
       @btime $A * $A;
  261.141 ns (2 allocations: 128 bytes)
```

At 46d66149bde3f651b80f607c1893e68332100b59:

```
julia> Base.GIT_VERSION_INFO.branch, Base.GIT_VERSION_INFO.commit
("matmuladd", "46d66149bde3f651b80f607c1893e68332100b59")

julia> using BenchmarkTools
       A = [1. 0; 2 0];
       @btime $A * $A;
  32.423 ns (1 allocation: 112 bytes)
```

tkf commented on stdlib/LinearAlgebra/src/generic.jl:
```diff
@@ -2,29 +2,33 @@
 end
 
-mul!(C::AbstractArray, s::Number, X::AbstractArray) = generic_mul!(C, X, s)
-mul!(C::AbstractArray, X::AbstractArray, s::Number) = generic_mul!(C, s, X)
```
FYI: I think there was a bug here. The order of `s` and `X` are flipped when calling `generic_mul!`.

tkf replied:
It was not a bug: https://github.com/JuliaLang/julia/issues/23919#issuecomment-441827558

tkf commented on stdlib/LinearAlgebra/test/generic.jl:
```diff
@@ -14,12 +14,16 @@ struct Quaternion{T<:Real} <: Number
 end
 Quaternion(s::Real, v1::Real, v2::Real, v3::Real) = Quaternion(promote(s, v1, v2, v3)...)
+Base.convert(::Type{Quaternion{T}}, s::Real) where {T <: Real} =
+    Quaternion{T}(convert(T, s), zero(T), zero(T), zero(T))
```
I had to add the definition of this `convert` and `+` below to do the test I mentioned in https://github.com/JuliaLang/julia/pull/29634#discussion_r235246958

tkf committed:
Annotate argument type MulAddMul

tkf committed:
Inline all addmul!

tkf committed:
Construct MulAddMul outside A_mul_B_td!

tkf commented:
Problem similar to https://github.com/JuliaLang/julia/pull/29634#issuecomment-440512432 happens with other combinations of matrix/vector types so I go ahead and make all `addmul!` inline in ce7b86260f281afe6969731e5e6b6c0797560bfa.

<details>
<summary>Benchmarks</summary>

`master`:

```julia
julia> Base.GIT_VERSION_INFO.branch, Base.GIT_VERSION_INFO.commit
("master", "ff74f6bc77e42c974b0f890ff25701f38f362560")

julia> using LinearAlgebra
       using BenchmarkTools
       n = 10
       A = Tridiagonal(collect.(Float64, (1:n-1, 1:n, 1:n-1))...);
       v = collect(1.0:n);
       @btime $A * $v;
  34.664 ns (1 allocation: 160 bytes)
```

Before ce7b86260f281afe6969731e5e6b6c0797560bfa:

```julia
julia> Base.GIT_VERSION_INFO.branch, Base.GIT_VERSION_INFO.commit
("matmuladd", "eb1e87fd260a7f6277a5a846fd1d89b02567f546")

julia> using LinearAlgebra
       using BenchmarkTools
       n = 10
       A = Tridiagonal(collect.(Float64, (1:n-1, 1:n, 1:n-1))...);
       v = collect(1.0:n);
       @btime $A * $v;
  262.449 ns (2 allocations: 176 bytes)
```

At ce7b86260f281afe6969731e5e6b6c0797560bfa:

```julia
julia> Base.GIT_VERSION_INFO.branch, Base.GIT_VERSION_INFO.commit
("matmuladd", "ce7b86260f281afe6969731e5e6b6c0797560bfa")

julia> using LinearAlgebra
       using BenchmarkTools
       n = 10
       A = Tridiagonal(collect.(Float64, (1:n-1, 1:n, 1:n-1))...);
       v = collect(1.0:n);
       @btime $A * $v;
  36.724 ns (1 allocation: 160 bytes)
```

</details>

tkf committed:
Add multiply-add interface in test/tridiag.jl

tkf commented:
I ran `BaseBenchmarks.SUITE["linalg"]` a few times and now it does not detect any regressions consistently (i.e., different result in different trials and cannot be reproduced in different machines).  This was the last TODO and there is nothing to add from me at the moment.  I appreciate it if someone familiar with this codebase let me know what I can do to get it merged.

tkf commented:
I forgot to mention that this is my first patch to Julia in a performance-critical section.  So please pay extra attention to possible performance bugs.

fredrikekre commented:
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/29634#issuecomment-440591703) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/d1335019410d2fad5a88347ca7346385203c6058/4ef7eb8_vs_464397e/report.md). cc @ararslan

tkf commented:
The only regressions detected in `linalg` are `sqrt` of triangular matrix wihch seems to not call `mul!`.  So I guess it's irrelevant?

BTW, can this PR get `linear algebra` tag and 1.1 milestone?

tkf committed:
Mention combined multiply-add in NEWS.md [ci skip]

tkf commented:
(@StefanKarpinski thanks for adding the tag!)

I added it in NEWS.md. I was not sure I should run `NEWS-update.jl` so I kept it as-is.  Running `NEWS-update.jl` produced the links I didn't add so I thought it was run occasionally.

tkf committed:
Mention that mul!(C, A, B, α, β) is deprecated [ci skip]

StefanKarpinski commented:
I've added "Pick a final name" to the checklist above so that someone doesn't go ahead and merge this thinking it's complete. Having but `addmul` and `muladd` seems... very confusing.

tkf commented:
Another TODO: use `rmul!` instead of `lmul!` (thanks @haampie!)

see:
https://github.com/JuliaLang/julia/issues/23919#issuecomment-441827558
https://github.com/JuliaLang/julia/blob/6b2d211d22288ef7e66fc0f5bd21d8db952f28ba/stdlib/LinearAlgebra/test/generic.jl#L219-L220
https://github.com/JuliaLang/julia/blob/58c0ed57dc796a9bb0759ce49781d97b8474eee3/stdlib/SparseArrays/src/linalg.jl#L39

tkf commented:
@andreasnoack, @haampie, (who else knows about it?)

What math definition of the API should we use?  @haampie pointed out in https://github.com/JuliaLang/julia/issues/23919#issuecomment-441827558 that it's decided(?) to be _C = ABα + Cβ_ but it looks like this is not strictly followed.

Currently, `mul!(C::StridedVecOrMat, A::SparseMatrixCSC, B::StridedVecOrMat, α::Number, β::Number)` implements _C = AαB + Cβ_:

https://github.com/JuliaLang/julia/blob/6bfe9476ff3861ec06ea1f547de8449cd184d13b/stdlib/SparseArrays/src/linalg.jl#L41-L48

whereas `mul!(C::StridedVecOrMat, adjA::Adjoint{<:Any,<:SparseMatrixCSC}, B::StridedVecOrMat, α::Number, β::Number)` implements _C = αAB + Cβ_:

https://github.com/JuliaLang/julia/blob/6bfe9476ff3861ec06ea1f547de8449cd184d13b/stdlib/SparseArrays/src/linalg.jl#L66-L74

This is the case for `Transpose`, too.

It seems _ABα_ is the only expression that can be implemented efficiently in the above two cases.  For symmetry, I guess it makes sense to use _Cβ_ for the other term.

So we should use _C = ABα + Cβ_ everywhere and fix SparseArrays, right?  I want to have a confirmation before fixing it.

andreasnoack commented:
Yes, please multiply the scalar from the right. Things have been implemented by different people at different times so things are not consistent but we work towards consistency.

Many of the conventions for how you write this are based on real or complex use cases where it doesn't matter. E.g. it is also standard to write the eigenvalue problem as `Av=λv` although the version with all the vectors is `AV=VΛ` so it would be better to write the first version as `Av=vλ`.

Anyway, would be great if you could update the PR to multiply scalars from the right. Nobody should be worse off and some of us will be happier.

tkf committed:
Change API definition to C = ABα + Cβ

tkf committed:
Fix indentation

tkf committed:
Merge branch 'master' into matmuladd

chriscoey commented:
Great! #30163 has been merged

amontoison commented:
Aside the choose for the final name, everything is fine for merging this pull request?
No news about it since the beginning of the year.

chriscoey commented:
now that JuliaCon is over, if no one has had any new ideas about this, maybe we can ask triage again - for an actual decision one way or the other?

tkf commented:
@amontoison, @chriscoey Thanks for the ping. Please see https://github.com/JuliaLang/julia/issues/23919#issuecomment-516667273

tkf committed:
Merge branch 'master' into matmuladd

tkf committed:
Fix triangular.jl

tkf committed:
Fix UndefRefError in multiplication with Diagonal

tkf committed:
Define Base.convert(::Type{Quaternion{T}}, s::Real)

This is required for testset "* and mul! for non-commutative scaling"
in LinearAlgebra/test/generic.jl to pass.

tkf committed:
Rename: addmul! -> mul!

tkf committed:
Fix doctest

tkf committed:
Workaround broadcast error with e.g., triangular matrix of BigFloat

tkf commented on stdlib/LinearAlgebra/test/addmul.jl:
```diff
@@ -0,0 +1,196 @@
+            Bc = Matrix(B)
+            @test collect(returned_mat) ≈ collect(α * Ac * Bc + β * Cc)  rtol=rtol
+        else
+            @test collect(returned_mat) ≈ collect(α * A * B + β * Cc)  rtol=rtol
```
ref #32883

tkf commented:
It passes all tests now (the failures in CI seems unrelated).  I think it's ready for review again.  I fixed all non-BLAS handling of beta=0 in the locations I noticed (though I may be missing something as the diff is large...)

tkf commented:
ping @andreasnoack

chethega commented on stdlib/LinearAlgebra/src/generic.jl:
```diff
@@ -2,29 +2,108 @@
+end
+
+MulAddMul(alpha::TA, beta::TB) where {TA, TB} =
+    MulAddMul{isone(alpha), iszero(beta), TA, TB}(alpha, beta)
```
It might be preferable for dispatch/inference to write this as
```
@inline function MulAddMul(alpha::TA, beta::TB) where {TA, TB}
    if isone(alpha)
        if iszero(beta)
           return MulAddMul{true, true, TA,TB}(alpha, beta)
       else
           return MulAddMul{true, false, TA, TB}(alpha, beta)
       end
       else
            if iszero(beta)
               return MulAddMul{false, true, TA,TB}(alpha, beta)
           else
                return MulAddMul{false, false, TA, TB}(alpha, beta)
           end
        end
       nothing
end
```
This would help the compiler/inference make an exhaustive list of cases instead of potentially open-ended dynamic dispatch.

tkf replied:
This is a very clever idea.  But, FYI, I think the benchmark I ran while ago https://github.com/JuliaLang/julia/pull/29634#issuecomment-440512432 suggests that constant propagation took care of it (at the cost of `@inline`ing functions).

chethega commented:
Can we get a nanosoldier to see potential perf regressions in the 3-arg case?

andreasnoack commented:
@nanosoldier `runbenchmarks("linalg", vs = ":master")`

Great work. Let's run see what Nanosoldier says.

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/29634#issuecomment-521379846) has completed - no performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/6242c5195076a67126ccb460ebb5f3b5b82909ef/09ef612_vs_57a022f/report.md). cc @ararslan

andreasnoack commented:
@ararslan Do you know why there are no results in the report?

ararslan commented:
It appears to have actually executed several groups of benchmarks: see [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/6242c5195076a67126ccb460ebb5f3b5b82909ef/09ef612_vs_57a022f/report.md#benchmark-group-list). The case in which they aren't listed is when there are no improvements or regressions, i.e. nothing is affected at all.

andreasnoack commented:
I'd forgotten that. It's quite rare that there are no changes detected at all.

Then let's get thing merged such that we can finally start using the feature. It's been in demand for quite some time.

JeffBezanson commented:
The tests for this are failing sometimes.

fredrikekre commented:
See #32924

Jutho commented:
Also, I notice that with a code which uses many small matrix multiplications, the construction of the `MulAddMul` object takes a significant fraction of the time. I first assumed this was an issue with profiling, but replacing `mul!` calls with corresponding `BLAS.gemm!` calls confirms that these profiles are telling the truth.

Here an example profile:
<img width="1222" alt="profile" src="https://user-images.githubusercontent.com/4162102/67866607-81b71100-fb29-11e9-8cc4-1fb0a86b6ec8.png">

tkf commented:
As for Adjoint/Transpose, I think we just need something like #33229 (as you've already discovered).

`MulAddMul` construction takes time when alpha and beta are not constant.  A MWE is

```julia
N = 10
A = randn(N, N)
B = randn(N, N)
C = similar(A)

function demo(C, A, B, alpha, beta, n)
    for _ in 1:n
        mul!(C, A, B, alpha, beta)
    end
end

demo(C, A, B, 1, 0, 1)
@time @profile demo(C, A, B, 1, 0, 10^5)
```

while `MulAddMul` vanishes from the profile if I do

```julia
function demo(C, A, B, n)
    for _ in 1:n
        mul!(C, A, B, 1, 0)
    end
end
```

Maybe something like https://github.com/JuliaLang/julia/pull/29634#discussion_r314010599 would be the solution?  Maybe we can also add an early branch in `mul!` to directly call `BLAS.gemm!` without constructing `MulAddMul`?

Jutho commented:
Thanks for the quick response and the pointers @tkf .

For the `MulAddMul` issue, I had constant values in the code, but maybe some higher levels up, so I don't know if they were propagated correctly. It also seemed to be worse for the `adjoint(A)*B` case than for the `A*B` case, which was maybe related to the first issue. I can profile further if needed, but have currently worked around these by directly calling `gemm!`, and it seems you are already aware of these issues. That's also why I didn't open new issues right away.

tkf commented:
Yeah, propagating constants for long call chains is not super robust... FYI, one option may be to use StaticNumbers.jl.  This calls `gemm!` and compiles away `MulAddMul` construction:

```julia
using StaticNumbers
demo(C, A, B, static(1), static(0), 1)
@time @profile demo(C, A, B, static(1), static(0), 10^5)
```

I also tried https://github.com/JuliaLang/julia/pull/29634#discussion_r314010599 and it works (i.e., `MulAddMul` vanishes from the profile). But, IIUC, it works by duplicating function body. Not sure if we want the code bloat, especially the ones implemented in pure Julia.
